ENH: Add Winter Values to Partition Explainer

CousinThrockmorton opened the pull request:
## Overview

Closes #3638 

Description of the changes proposed in this pull request:

Extend SHAP to calculate feature contributions for coalitions of features Owen and Winter values. Having reviewed the Partition Explainer and Permutation Explainer code I am not aware of any current package or method that allows users to calculate these values with context specific coalition structures.

The current PartitionExplainer relies on the scipy.cluster.hierarchy package to simplify the permutations required by Shapley value by restricting the calls to those consistent with the coalitions based on the closeness of feature. Therefore, it is not as stated the Owen value but the Winter Value for the hierarchical clustering.  

For the computational strategy please refer to [Values for level structures with polynomial-time algorithms,
relevant coalition functions, and general considerations](https://www.sciencedirect.com/science/article/pii/S0166218X21004637)

## Checklist

- [x] Adopt utils.masked_models to accept custom non-binary hierarchies(/ translate them to binary trees and create masks of them at the specified level. (isn't possible and computationally an inferior approach))
- [x] Simplify explainer._partition.py code for faster execution and clarify code documentation.

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)

CloseChoice commented:
I am looking into this PR right now. It's a bit hard to actually see the difference/similarities to the existing implementation, will try to change a few things there. Could you please provide a link without a pay wall to the winter value formula? I am not really familiar with the concept and didn't find any definition in the papers you linked in the issue.

EDIT: is there a test case we can implement? We need this for new functionality.

CousinThrockmorton commented:
https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=90811fdcda3f3243fa57c0e2d14926a70e5858b1 

Here is the link without a paywall, the description of the value can be found on pages 22-23 describing Winter (1989). I am sorry for the slow progress I had been going over the current implementation and a bit new to this. 

I believe the PartitionExplainer and the ExactExplainer with the clustering option both calculate the Winter value not the Owen value (only respecting flat clusters) as stated in the documentations as the permutations used respect the whole clustering tree created by scipy. I believe the logic behind this approximation is that the marginals for correlated features will be similar with respect to the other correlated features and therefore we can reduce the number of permutations. 

The point I am stuck at is in order to pass user described hierarchies the scipy binary hierarchy is not sufficient. However I am not sure what would be the most user friendly way to define the hierarchy so it can be checked and used easily. I will base my approach on the ExactExplainer as with the clustering option it is faster than the PartitionExplainer and write a new make_masks function (the methods mainly differ in how they use the results of this function to get to the same point) the rest should be the same. Do you think this option should be added to the ExactExplainer or the Partition method?  

For example the exact method calculates for features [1,2,3,4] the Shapley value for 1.

[1] - []
[1,2] -[2]
[1,3] -[3]
[1,4] -[4]
[1,2,3] -[2,3]
[1,3,4] -[3,4]
[1,2,4] -[2,4]
[1,2,3,4] -[2,3,4]

The Owen value would be with clustering [[1,2,],[3,4]].

[1] - []
[1,2] -[2]
[1,3,4] -[3,4]
[1,2,3,4] -[2,3,4]

The Winter value [[1,2],[[3,4],[5]]]. I hope this structure makes the hierarchy clear, essentially we are talking of nested partitions of features. 

[1] - []
[1,2] -[2]
[1,3,4,5] -[3,4,5]
[1,2,3,4,5] -[2,3,4,5]

So the features marginal contribution is to the cluster that it is part of and to the clusters it is external to. And this is exactly what the current method does with the binary clustering trees. For the notebook on PartitionExplainer this is the clustering used.

![image](https://github.com/shap/shap/assets/84054773/364b830c-1f23-4f3a-b236-dbe8cf1020bb)

For the feature HouseAge the following masks are used which is what the Winter Value specifies. 

X.columns = Index(['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup',
       'Latitude', 'Longitude'],

(marginal to the external tree) [12] 
[ True, False, False, False,  True,  True, False,  True],  
[ True,  True, False, False,  True,  True, False,  True],
(marginal with all turned on) [12, 10]
[ True, False,  True,  True,  True,  True,  True,  True],
[ True,  True,  True,  True,  True,  True,  True,  True],
(marginal with all off) []
[False, False, False, False, False, False, False, False], 
[False,  True, False, False, False, False, False, False],
(marginal to the internal tree) [10]
[False, False,  True,  True, False, False,  True, False],
[False,  True,  True,  True, False, False,  True, False],

or for Latitude 

External [12 ,1]
[ True  True False False  True  True False  True]
[ True  True False False  True  True  True  True]
All turned on [12 ,1 ,8]
[ True  True  True  True  True  True False  True]
[ True  True  True  True  True  True  True  True]
External tree  [12]
[ True False False False  True  True False  True]
[ True False False False  True  True  True  True]
All but HouseAge [12 ,8]
[ True False  True  True  True  True False  True]
[ True False  True  True  True  True  True  True]
To the HouseAge [1]
[False  True False False False False False False]
[False  True False False False False  True False]
Internal [1 ,8]
[False  True  True  True False False False False]
[False  True  True  True False False  True False]
All turned off []
[False False False False False False False False]
[False False False False False False  True False]
Local cluster [8]
[False False  True  True False False False False]
[False False  True  True False False  True False]


A couple examples of hierarchies I would like to pass [[1,2],[3,4,5]] or [[1,2],[[3,4][5,6]]] which can not be necessarily translated to a scipy hierarchy. I hope this clarifies what I would like to do as well as the current state of the ExactExplainer and PartitionExplainer methods. I hope you can give me some guidance on how to proceed. I really appreciate your help apologies for the long reply. Hope you have a good day!

CloseChoice commented:
Thanks, this helps a lot. Will take a deeper look in the next week and we can follow this up here.

CousinThrockmorton commented:
Thanks for the help. You can find most of my progress in the Non-binary_tree_masking notebook under notebooks. I am trying to figure out how to generate all combinations of masks of top node siblings and local siblings.  This is done in the current partition method via traversing the tree and adding to each nodes mask its siblings mask on the left and right, then traversing the courser partition on the left and right and going lower if possible. See my comment before. This way we end up with all possible combinations top siblings and own path siblings in line with the Winter value calculation. 

Now this is a slightly harder with n-ary trees as there is no easy left and right.  At each level we need to calculate all possible combinations of masks then refine all these combinations at the lower level but one by one?

CloseChoice commented:
The paper without a paywall doesn't work anymore. I read [Winter (1989](https://link.springer.com/article/10.1007/BF01268161) but did not find a concrete formula that can be implemented in a straight forward fashion. Could you point me to that?

Also I would think that we are missing some combinations here:
> The Winter value [[1,2],[[3,4],[5]]]. I hope this structure makes the hierarchy clear, essentially we are talking of nested partitions of features.
[1] - []
[1,2] -[2]
[1,3,4,5] -[3,4,5]
[1,2,3,4,5] -[2,3,4,5]

Aren't we missing:
[1, 5] - [5]
[1, 3, 4] - [3, 4]
?

CousinThrockmorton commented:
Hey, thanks for looking into the paper. Here is an other publication that interprets the use of the partition_explainer. https://towardsdatascience.com/shaps-partition-explainer-for-language-models-ec2e7a6c1b77. The formula is the same, the average of marginals, as Shapley values. In fact there is an interpretation that Winter values are a recursive call of Shapley values with respect to the coalitions. 

That is for the coalitions structure [[1,2],[[3,4],[5]]. One way to get the Winter values is to consider [1,2], and [[3,4],[5]] as the players and calculate their Shapley values getting a value for each cluster. Then use the assigned value for each cluster as the payoff to calculate Shapley values on the constituent parts of each cluster for instance between  [3,4], and [5]. And so on and so forth. 

This recursive method could work. However, I believe the direct option may be more straight forward.  The idea is to consider external and internal coalitions to get the value. External coalitions are that of the top level assigning the feature value with respect to other coalitions. Then the internal coalitions assign the value in group to the feature. So the answer to your question is, no as  [[3,4],[5]] are external to the cluster of [1,2] and form one cluster.

CousinThrockmorton commented:
Heyoo, 

I have made one version of the method work and reproduce the binary methods results consistently. Please see the [Rewriting.ipynb](https://github.com/CousinThrockmorton/shap/blob/master/notebooks/Rewriting.ipynb) notebook. There are still plenty of issues, in specific how the methods are selected, the mask_shape output for multi output (?), possibly weighted average over permutations, and iterative strategy to allow for early stopping which should be possible with the n-ary method as well. The code itself could also use some cleaning up to get rid of a few redundant operations, which I will do later in July as well as maybe add a visualization module that plots the combined features. 

The basic logic is the following:

1. Use an n-ary tree structure (possibly can be streamlined via a package, or first node, siblings formulation)
2. Assign every node all possible permutations of sibling nodes. 
3. Depth first traversal of the tree collecting all permutations on path and creating all possible combinations of permutations
4. Create a mask list for every node that is the length of all leafs and encodes all leafs below a node
5. Combine the masks along the combinations of permutations. This will be the operations to calculate the Winter values.
6. Call the model on the unique mask combinations and calculate the average marginals as usual. 

I hope this helps. Let me know what you think.

CousinThrockmorton commented:
[Values for level structures with polynomial-time algorithms,
relevant coalition functions, and general considerations](https://www.sciencedirect.com/science/article/pii/S0166218X21004637)

This may be the best literature for the computation of such value. I am yet to implement this and my current code doesn't manage memory efficiently when computing the products of the power sets needed to get all permutations consistent with the levels structure. See [League of Legends Win Prediction with XGBoost.ipynb](https://github.com/CousinThrockmorton/shap/blob/master/notebooks/tabular_examples/tree_based_models/League%20of%20Legends%20Win%20Prediction%20with%20XGBoost.ipynb) which is the largest tabular example I had been able to find. The code fails in specific due to the formulation of the generate_paths_and_combinations function using itertools.product on all the on path power sets of sibling nodes. 

By the way. I am writing a thesis on this extension project that I hope to share in some format as I have seen several recent requests on the academic documentation of the PartitionExplainer.

CloseChoice commented:
> [Values for level structures with polynomial-time algorithms, relevant coalition functions, and general considerations](https://www.sciencedirect.com/science/article/pii/S0166218X21004637)
> 
> This may be the best literature for the computation of such value. I am yet to implement this and my current code doesn't manage memory efficiently when computing the products of the power sets needed to get all permutations consistent with the levels structure. See [League of Legends Win Prediction with XGBoost.ipynb](https://github.com/CousinThrockmorton/shap/blob/master/notebooks/tabular_examples/tree_based_models/League%20of%20Legends%20Win%20Prediction%20with%20XGBoost.ipynb) which is the largest tabular example I had been able to find. The code fails in specific due to the formulation of the generate_paths_and_combinations function using itertools.product on all the on path power sets of sibling nodes.
> 
> By the way. I am writing a thesis on this extension project that I hope to share in some format as I have seen several recent requests on the academic documentation of the PartitionExplainer.

Thanks for all of the effort. I think I understand the problem a bit better now, may I suggest a course of action (this is basically how I would approach this, but feel free to follow a different path):
 - let's not care about performance for now
 - implementing a solution that does not cover all edge cases is fine for me, as long as we raise insightful error messages in the uncovered cases
 - I would focus on tests for now, we can tune performance and improve generality once something is there. I do not see the need to have a perfect solution from the start.
 - Concretely: feel free to overwrite the PartitionExplainer in `_partition.py`, then we can see the tests that do not work as of now, and can add new test cases for the winter values. Before merging all test cases must pass, of course. I checked your code and in case of the PartitionExplainer you just added comments, so I think it'll be safe to overwrite and just adding the new masker functionality.

If I understood correctly then implementing the winter values means to "just" mask the inputs correctly, I didn't yet quite understand how, but will invest a bit more time into the specifics in the future.

Thanks again so much, I think this is a great opportunity to add functionality (and to cleanup the PartitionExplainer, but that'll be on me). I won't be available over the next 2 weeks though.

CousinThrockmorton commented:
Hope you enjoyed your time away. I took the time to write up a thesis work on the topic I hope it helps to illustrate the working of the method and the logic behind the explanations. 
[Winter_Values_.pdf](https://github.com/user-attachments/files/16248121/Winter_Values_.pdf)
I will soon get to adopting the code so it passes the tests and works top down.

abarbosa94 commented:
Hey @CousinThrockmorton, I hope you're doing well :) I just found this thread and got interested in your work. Congrats on that. :) I'm working on a master's thesis on applying Owen values to specific text data structures. I think this application of Winter values could be interesting to my work. As you closed this PR, I was wondering if you still have plans to work on such integration to the main shap library

Please let me know if you need/want any help on that :)

Thanks!

CousinThrockmorton commented:
Heyoo, @abarbosa94,  Thank you for your interest. Yes, I had been working on a better calculation strategy that could deal with large dimensional models such as text models, but I have made a little mess of my fork in the meantime.

CousinThrockmorton commented:
@CloseChoice I had resolved all merge conflicts, all pre-commit checks, ruff checks, and pytests pass. Could you approve the workflows for this PR?

I had not written a new test case yet to be added to the partition_test cases, one idea would be to check the two methods against each other, as illustrated by the [Simple California Demo](https://github.com/CousinThrockmorton/shap/blob/master/notebooks/tabular_examples/model_agnostic/Simple%20California%20Demo.ipynb). That is providing the clustering used by the exact or the clustering method as a partition_tree should result in the same result as those two methods. As seen in this visualisation. 
![output](https://github.com/user-attachments/assets/559503e8-c809-49bd-9b09-3db13fbe63cb)

CousinThrockmorton commented:
Hello @CloseChoice,

I had been busy with some personal issues. But I finally came back to this. I split off the new function I wrote to [_coalition.py](https://github.com/CousinThrockmorton/shap/blob/master/shap/explainers/_coalition.py) as even though it extends the functionality I doubt it can be as efficient as the partition method so I will leave the _partition.py as is. 

I also wrote some tests in [test_coalition.py ](https://github.com/CousinThrockmorton/shap/blob/master/tests/explainers/test_coalition.py)that check that the coalition method matches up with the exact explainer, partition explainer and that the values sum correctly. 

I also rewrote the [Simple California Demo](https://github.com/CousinThrockmorton/shap/blob/master/notebooks/tabular_examples/model_agnostic/Simple%20California%20Demo.ipynb) to showcase the different methods that use different forms of coalitional cooperative game theory. 

There is a lot of work left. First, multi output doesn't work. Second, the evaluation is very inefficient and I believe it can be improved via a top down evaluation method via a concurrent.futures traversal. Third, I hope to write some more demonstrations, first for temporal analysis, then using a meronomy dataset or documentation for textual models. 

Hope this should get this project moving forward. Sorry for the delays.

CloseChoice commented:
> Hello @CloseChoice,
> 
> I had been busy with some personal issue of homelessness. But I finally came back to this. I split off the new function I wrote to [_coalition.py](https://github.com/CousinThrockmorton/shap/blob/master/shap/explainers/_coalition.py) as even though it extends the functionality I doubt it can be as efficient as the partition method so I will leave the _partition.py as is.
> 
> I also wrote some tests in [test_coalition.py ](https://github.com/CousinThrockmorton/shap/blob/master/tests/explainers/test_coalition.py)that check that the coalition method matches up with the exact explainer, partition explainer and that the values sum correctly.
> 
> I also rewrote the [Simple California Demo](https://github.com/CousinThrockmorton/shap/blob/master/notebooks/tabular_examples/model_agnostic/Simple%20California%20Demo.ipynb) to showcase the different methods that use different forms of coalitional cooperative game theory.
> 
> There is a lot of work left. First, multi output doesn't work. Second, the evaluation is very inefficient and I believe it can be improved via a top down evaluation method via a concurrent.futures traversal. Third, I hope to write some more demonstrations, first for temporal analysis, then using a meronomy dataset or documentation for textual models.
> 
> Hope this should get this project moving forward. Sorry for the delays.

Hi @CousinThrockmorton, no worries about the delay and thanks for the update. I really hope you are okay.

I won't stop you from continuing to work on this but I would suggest that we first make this ready and then we can improve to multi-output and also improve performance. For now we can even just check whether we are in the multi-output case and throw a `NotImplementedError` in this case. But if you want to finish up things first, I also support that. Whatever you prefer here.

CousinThrockmorton commented:
Hi @CloseChoice,

Fixed the multioutput functionality, and added the test as recommended so no need for errors. Thanks for the advice and concern. I hope this allow you to move forward with introducing the functionality. Let me know what else should I do to enable merging.

I suppose then to move step by step. I can open a new PR on enhancing the efficiency of the method, making sure it works with different data modalities, and possibly visualisations like in the California demo? I would appreciate any ideas you might have for cool demos or uses.

CousinThrockmorton commented:
Hey @CloseChoice 

Could you check on my branch. Any specific next steps you would suggest?

CloseChoice commented on notebooks/tabular_examples/tree_based_models/winter_values_values.npy:
I would suggest getting rid of all the .npy files, IMO we can store data in those in a notebook but pls not check them in. Also if you generate one of these in a notebook, pls add `.npy` files to the .gitignore

CloseChoice commented on notebooks/tabular_examples/tree_based_models/League of Legends Win Prediction with XGBoost.ipynb:
I assume you just reexecuted these notebooks. Could you pls do

`git checkout master <notebook>` for all notebooks that you didn't touch. So that we don't have any unnecessary changes.

CloseChoice commented on shap/actions/_action.py:
```diff
@@ -1,14 +1,6 @@
 class Action:
     """Abstract action class."""
 
-    def __init__(self, cost):
```
why remove these?

CloseChoice commented on shap/actions/_optimizer.py:
```diff
@@ -9,17 +7,17 @@
 
 class ActionOptimizer:
-    def __init__(self, model, actions: list[Action | list[Action]]):
+    def __init__(self, model, actions):
```
I assume these changes are since you are not working on the most recent master branch or something got messed up while resolving merge conflicts? Or is there a particular reason to change this?

CloseChoice reviewed (changes requested):
@CousinThrockmorton thanks for coming back to this. I think the best way to tackle this is to first trim the PR down to only the changes that you've done and resolving all junk that results from merge conflicts. Let me know if you need help here, then I would suggest that we'll either a quick meeting to sort it out together or you name the files which you touched and I can give it a try.

CousinThrockmorton committed:
Add CoalitionExplainer, test case, and two example notebooks discrading all other changes

CousinThrockmorton commented:
Hey @CloseChoice I just pulled the latest shap and added the files I have touched to clean up the PR. Seemed like the easiest solution, keeping my dev mess, might not be the most elegant, hihi. Let me know if you were thinking of some other solution. All tests and pre-commits pass for me.  I am available any time if you would like to have a short meeting for integration. Thanks for dealing with my mess :-)

CousinThrockmorton committed:
Notebook execution fix

CloseChoice commented on shap/explainers/_coalition.py:
```diff
@@ -0,0 +1,512 @@
+    Textual and image data is not yet implemented.
+
+    The explainer supports both single and multi-output models, and can handle various
+    types of input data through the provided masker.
```
is there a specific paper we can quote?

CousinThrockmorton replied:
Not, yet. I am writing something more publishable. For now, the best reference may be my masters thesis posted on the PR.

CloseChoice replied:
Do you feel comfortable to add this here for reference? Of course you can come back at any time and replace it with something else. If not, then we can leave it as is and you'll just come back and file another PR once it is ready.

CloseChoice commented on shap/explainers/_coalition.py:
```diff
@@ -0,0 +1,512 @@
+
+        if len(call_args) > 0:
+
+            class CoalitionExplainer(self.__class__):
```
Could you pls elaborate a bit on the idea behind this class? Why don't we just add **kwargs to the __call__ method?

CousinThrockmorton replied:
The "head" of the class I just took from the partition explainer. There they point to being able to dynamically modify the call at runtime: 
"
        # if we have gotten default arguments for the call function we need to wrap ourselves in a new class that
        # has a call function with those new default arguments
"

CloseChoice replied:
but why do they need to be default arguments? Can't we just pass them to __call__ via **kwargs? Or could you pls show a use case where we actually need that?

CloseChoice commented on shap/explainers/_coalition.py:
```diff
@@ -0,0 +1,512 @@
+        keys.append(node.key)
+
+        for subset in node.child:
+            child_masks, child_keys = create_masks1(subset, columns)
```
so we have create_masks and create_masks1. Is there a way to add a more expressive name?

CousinThrockmorton replied:
That is just dev residue, there isn't a create_masks I just forgot to rename it.

CloseChoice commented on shap/explainers/_coalition.py:
```diff
@@ -0,0 +1,512 @@
+                if key in masks_dict:
+                    masks.append(masks_dict[key])
+
+        if masks:
```
NIT, pls do `if len(masks) > 0` which is more expressive.

CousinThrockmorton replied:
Agreed. Done

CloseChoice commented on shap/maskers/_tabular.py:
```diff
@@ -66,7 +67,21 @@ def __init__(self, data, max_samples=100, clustering=None):
 
         # compute the clustering of the data
-        if clustering is not None:
+        # if clustering is not None:
```
can we remove this?

CousinThrockmorton replied:
Deleted the commented out old implementation.

CloseChoice commented on shap/explainers/_coalition.py:
```diff
@@ -0,0 +1,512 @@
+    return combined_masks
+
+
+def map_combinations_to_unique_masks(combined_masks, unique_masks):
```
these are all specific functions to the CoalitionExplainer, can we therefore indicate them as private by prefixing with `_`?

CousinThrockmorton replied:
Done.

CloseChoice commented on tests/explainers/test_coalition.py:
```diff
@@ -0,0 +1,82 @@
+    )
+
+
+def test_tabular_coalition_exact_match():
```
very nice test, thanks!

CousinThrockmorton replied:
Thank you :-)

CloseChoice reviewed (changes requested):
looks amazing, just a couple nitpicks. Let me know if you need help with the pipelines.

CousinThrockmorton committed:
fixing review comments

CousinThrockmorton committed:
fixing review comments, minor fixes

CousinThrockmorton commented:
Hey hey @CloseChoice, thank you for the review. I replied to the comments and applied the suggested changes. Regarding the pipelines, I am not sure about how to resolve the mypy issues, I removed the superfluous package from the notebook so that check should pass now. Thanks for all the support, I am learning a lot. Let me know anything I could do. :-)

CloseChoice commented:
> Hey hey @CloseChoice, thank you for the review. I replied to the comments and applied the suggested changes. Regarding the pipelines, I am not sure about how to resolve the mypy issues, I removed the superfluous package from the notebook so that check should pass now. Thanks for all the support, I am learning a lot. Let me know anything I could do. :-)

Alright, this looks really good, the test coverage is also great. Leave the mypy issue to me, I'll be doing this sometime this week. We basically have just one point to disucss then it's good to merge. Thanks already, this is absolutely amazing work.

CloseChoice committed:
fix mypy errors

CloseChoice committed:
Merge branch 'master' into FIX-mypy

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CloseChoice committed:
fix mypy errors

CloseChoice committed:
Merge branch 'FIX-mypy' of https://github.com/shap/shap into FIX-mypy

CloseChoice committed:
fix mypy errors

CloseChoice commented:
@CousinThrockmorton created a PR in your branch (https://github.com/CousinThrockmorton/shap/pull/1), pls check that, comment and if that is in, we are good to merge. That fixes the mypy issues, some left over comments in the notebooks, and also removes the dynamical creation of classes (though we can reintroduce that).

CousinThrockmorton committed:
Merge pull request #1 from CousinThrockmorton/FIX-mypy

Fix mypy

CloseChoice reviewed (approved):
LGTM, thanks a lot for this major effort. Special thanks for adding notebooks, tests and the like. 

The docs pipeline fails right now, but as far as I see that was a disconnect issue, therefore I'll give it a go and in case this is still breaking will file another PR.

IMW112 commented:
I just returned to the project I had been working on when I posted this [feature request](https://github.com/shap/shap/issues/3397)

Could barely believe my good fortune that someone has provided the solution in the meantime. so just saying thanks to @CousinThrockmorton for all your hard work on this. Looking forward to trying it out.
